File: regular_assignments/day_009/subarray_with_given_sum_and_length/solution.go
package subArrayWithGiveSumAndLength

func Call() {
	solve([]int{4, 3, 2, 6, 1}, 3, 11)
	solve([]int{4, 3, 2, 6, 1}, 4, 6)
}

func solve(arr []int, expectedLen, expectedSum int) int {
	prefixSumArr := createPrefixSum(arr)
	exists := 0
	i := 0
	j := expectedLen - 1
	for j < len(arr) {
		sum := 0
		if i == 0 {
			sum = prefixSumArr[j]
		} else {
			sum = prefixSumArr[j] - prefixSumArr[i-1]
		}
		if sum == expectedSum {
			exists = 1
		}
		i++
		j++
	}
	return exists
}

func createPrefixSum(arr []int) []int {
	prefix_sum := []int{arr[0]}
	for i := 1; i < len(arr); i++ {
		prefix_sum = append(prefix_sum, prefix_sum[i-1]+arr[i])
	}
	return prefix_sum
}

File: regular_assignments/day_009/subarray_with_given_sum_and_length/solution.rb
class Solution
  def call
    solve([4, 3, 2, 6, 1], 3, 11) # 1
    solve([4, 2, 2, 5, 1], 4, 6) # 0
  end

  def solve(arr, expected_len, expected_sum)
    prefix_sum_arr = create_prefix_sum_arr(arr)
    exists = 0
    i = 0
    j = expected_len-1
    while j < arr.length
      sum = 0
      sum = if i.zero?
              prefix_sum_arr[j]
            else
              prefix_sum_arr[j] - prefix_sum_arr[i - 1]
            end
      exists = 1 if sum == expected_sum
      i += 1
      j += 1
    end
    exists
  end

  def create_prefix_sum_arr(arr)
    prefix_sum_arr = []
    i = 0
    while i < arr.length
      if i.zero?
        prefix_sum_arr[i] = arr[i]
      else
        prefix_sum_arr[i] = prefix_sum_arr[i - 1] + arr[i]
      end
      i += 1
    end
    prefix_sum_arr
  end
end

# puts '-' * 100
# puts Solution.new.solve([4, 3, 2, 6, 1], 3, 11)
# puts '-' * 100
# puts Solution.new.solve([4, 2, 2, 5, 1], 4, 6)
